src/gsym/parser.rs

```diff
@@ -0,0 +1,330 @@
+        AddressInfo {
+            size,
+            name,
+            data: &self.raw_data[ad_off + 8..],
+        }
+    }
+
+    /// Get the string at the given offset from the String Table.
+    ///
+    /// # Saftety
+    ///
+    /// The code will crash with an invalid offset.
+    #[inline(always)]
+    pub fn get_str(&self, off: usize) -> &str {
+        assert!(self.str_tab[self.str_tab.len() - 1] == 0);
+        assert!(off < self.str_tab.len());
+        unsafe {
+            CStr::from_ptr((&self.str_tab[off..]).as_ptr() as *const i8)
+                .to_str()
+                .unwrap()
+        }
+    }
+}
+
+/// Find the index of an entry in the address table most likely
+/// containing the given address.
+///
+/// The callers should check the respective `AddressInfo` to make sure
+/// it is what they request for.
+pub fn find_address(ctx: &GsymContext, addr: u64) -> usize {
+    let mut left = 0;
+    let mut right = ctx.num_addresses();
+
+    if right == 0 {
+        return 0;
+    }
+    if addr < ctx.addr_at(0) {
+        return 0;
+    }
+
+    while (left + 1) < right {
+        let v = (left + right) / 2;
+        let cur_addr = ctx.addr_at(v);
+
+        if addr == cur_addr {
+            return v;
+        }
+        if addr < cur_addr {
+            right = v;
+        } else {
+            left = v;
+        }
+    }
+    left
+}
+
+/// Parse AddressData.
+///
+/// AddressDatas are items following AndressInfo.
+/// [`GsymContext::addr_info()`] returns the raw data of AddressDatas as a
+/// slice at [`AddressInfo::data`].
+///
+/// # Arguments
+///
+/// * `data` - is the slice from AddressInfo::data.
+///
+/// Returns a vector of [`AddressData`].
+pub fn parse_address_data(data: &[u8]) -> Vec<AddressData> {
+    let mut data_objs = vec![];
+
+    let mut off = 0;
+    while off < data.len() {
+        let typ = decode_uword(&data[off..]);
+        off += 4;
+        let length = decode_uword(&data[off..]);
+        off += 4;
+        let d = &data[off..(off + length as usize)];
+        data_objs.push(AddressData {
+            typ,
+            length,
+            data: d,
+        });
+        off += length as usize;
+
+        #[allow(non_upper_case_globals)]
+        match typ {
+            InfoTypeEndOfList => {
+                break;
+            }
+            InfoTypeLineTableInfo | InfoTypeInlineInfo => {}
+            _ => {
+                eprintln!("unknown info type");
+            }
+        }
+    }
+
+    data_objs
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::env;
+    use std::fs::File;
+    use std::io::Read;
+    use std::path::Path;
+
+    #[test]
+    fn test_parse_context() {
+        let args: Vec<String> = env::args().collect();
+        let bin_name = &args[0];
+        // data/test.gsym is generated by running gsymutil on the test
+        // case runner at target/debug/deps/blazesym-xxxx, which is
+        // generated by cargo for BlazeSym.
```

Comment:
In my opinion, we should come up with a more stable binary (more minimal Rust one or perhaps even C, as it tends to have more well defined and stable ABI) and use that as a base as part of the build step. Checking in a 4 MiB binary blob is not great. Having it depend on various Rust specific internals such as file paths and exact symbol names strikes me as very fragile.